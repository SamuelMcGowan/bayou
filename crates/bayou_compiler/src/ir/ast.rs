use bayou_diagnostic::span::Span;

use super::ir::Type;
use super::{BinOp, Ident, Node, UnOp};

#[derive(Node!)]
pub struct Module {
    pub items: Vec<Item>,
}

#[derive(Node!)]
pub enum Item {
    FuncDecl(FuncDecl),
    ParseError,
}

#[derive(Node!)]
pub struct FuncDecl {
    pub name: Ident,
    pub ret_ty: Type,
    pub statements: Vec<Stmt>,
}

#[derive(Node!)]
pub enum Stmt {
    Assign { ident: Ident, expr: Expr },

    Return(Expr),

    ParseError,
}

#[derive(Node!)]
pub struct Expr {
    pub kind: ExprKind,
    pub span: Span,
}

impl Expr {
    pub fn new(kind: ExprKind, span: Span) -> Self {
        Self { kind, span }
    }
}

#[derive(Node!)]
pub enum ExprKind {
    Constant(i64),

    Var(Ident),

    UnOp {
        op: UnOp,
        expr: Box<Expr>,
    },

    BinOp {
        op: BinOp,
        lhs: Box<Expr>,
        rhs: Box<Expr>,
    },

    Void,

    ParseError,
}
