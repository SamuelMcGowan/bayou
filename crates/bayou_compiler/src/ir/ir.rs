use super::{BinOp, Ident, Node, NodeCopy, UnOp};
use crate::symbols::LocalId;

pub struct FuncDecl {
    pub name: Ident,
    pub statements: Vec<Stmt>,
}

#[derive(Node!)]
pub enum Stmt {
    Assign { local: LocalId, expr: Expr },
    Return(Expr),
}

#[derive(Node!)]
pub struct Expr {
    pub kind: ExprKind,
    pub ty: Type,
}

#[derive(Node!)]
pub enum ExprKind {
    Constant(Constant),
    Var(LocalId),
    UnOp {
        op: UnOp,
        expr: Box<Expr>,
    },
    BinOp {
        op: BinOp,
        lhs: Box<Expr>,
        rhs: Box<Expr>,
    },
}

#[derive(NodeCopy!)]
pub enum Constant {
    I64(i64),
}

impl Constant {
    pub fn ty(&self) -> Type {
        match self {
            Self::I64(_) => Type::I64,
        }
    }
}

#[derive(NodeCopy!)]
pub enum Type {
    I64,
}
