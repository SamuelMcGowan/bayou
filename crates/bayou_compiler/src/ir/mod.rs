use bayou_diagnostic::span::Span;

pub mod ast;
#[allow(clippy::module_inception)]
pub mod ir;
pub mod token;

pub type InternedStr = lasso::Spur;
pub type Interner = lasso::Rodeo;

derive_alias! {
    #[derive(Node!)] = #[derive(Debug, Clone, PartialEq, Eq, serde::Serialize)];
    #[derive(NodeCopy!)] = #[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, serde::Serialize)];
}

#[derive(Node!, Copy)]
pub struct Ident {
    pub ident: InternedStr,
    pub span: Span,
}

#[derive(NodeCopy!)]
pub enum BinOp {
    Add,
    Sub,
    Mul,
    Div,
    Mod,

    BitwiseAnd,
    BitwiseOr,
    BitwiseXor,
    // Eq,
    // NotEq,

    // Gt,
    // Lt,
    // GtEq,
    // LtEq,
}

#[derive(NodeCopy!)]
pub enum UnOp {
    Negate,
    BitwiseInvert,
}

#[derive(NodeCopy!)]
pub struct Spanned<T> {
    pub node: T,
    pub span: Span,
}

impl<T> Spanned<T> {
    pub fn new(node: T, span: Span) -> Self {
        Self { node, span }
    }
}

impl<T, E> Spanned<Result<T, E>> {
    pub fn transpose(self) -> Result<Spanned<T>, E> {
        self.node.map(|node| Spanned {
            node,
            span: self.span,
        })
    }
}
