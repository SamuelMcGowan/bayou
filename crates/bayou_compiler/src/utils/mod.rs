pub mod keyvec;
pub mod peek;

macro_rules! assert_yaml_snapshot_with_source {
    ($source:expr => $output:expr) => {{
        insta::with_settings!({
            info => &$source,
            omit_expression => true,
        }, {
            insta::assert_yaml_snapshot!($output);
        })
    }};
}
pub(crate) use assert_yaml_snapshot_with_source;
