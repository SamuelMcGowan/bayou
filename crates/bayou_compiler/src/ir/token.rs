use bayou_diagnostic::span::Span;

use super::NodeCopy;
use crate::ir::InternedStr;

#[derive(NodeCopy!)]
pub struct Token {
    pub kind: TokenKind,
    pub span: Span,
}

#[derive(NodeCopy!)]
pub enum TokenKind {
    Keyword(Keyword),
    Identifier(InternedStr),
    Integer(i64),

    LBrace,
    RBrace,
    LParen,
    RParen,
    Semicolon,
    Bang,
    Arrow,

    Add,
    Sub,
    Mul,
    Div,
    Mod,

    Assign,
    // TODO: uncomment when assignment is added
    // AddEq,
    // SubEq,
    // MulEq,
    // DivEq,
    // ModEq,

    // TODO: uncomment when boolean operations are added
    // EqEq,
    // NotEq,

    // Gt,
    // Lt,
    // GtEq,
    // LtEq,
    BitwiseAnd,
    BitwiseOr,
    BitwiseXor,
    BitwiseInvert,
}

#[derive(NodeCopy!)]
pub enum Keyword {
    Func,
    Return,

    Let,

    Void,
    I64,
}

impl TokenKind {
    pub fn token_name(&self) -> &'static str {
        match self {
            TokenKind::Keyword(kw) => match kw {
                Keyword::Func => "keyword `func`",
                Keyword::Return => "keyword `return`",
                Keyword::Let => "keyword `let`",
                Keyword::I64 => "keyword `i64`",
                Keyword::Void => "keyword `void`",
            },
            TokenKind::Identifier(_) => "identifier",
            TokenKind::Integer(_) => "integer",
            TokenKind::LBrace => "`{`",
            TokenKind::RBrace => "`}`",
            TokenKind::LParen => "`(`",
            TokenKind::RParen => "`)`",
            TokenKind::Semicolon => "`;`",
            TokenKind::Bang => "`!`",
            TokenKind::Arrow => "`->`",
            TokenKind::Add => "`+`",
            TokenKind::Sub => "`-`",
            TokenKind::Mul => "`*`",
            TokenKind::Div => "`/`",
            TokenKind::Mod => "`%`",
            TokenKind::Assign => "`=`",
            TokenKind::BitwiseAnd => "`&`",
            TokenKind::BitwiseOr => "`|`",
            TokenKind::BitwiseXor => "`^`",
            TokenKind::BitwiseInvert => "`~`",
        }
    }
}
