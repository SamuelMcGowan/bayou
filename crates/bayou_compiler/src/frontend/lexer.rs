use std::str::Chars;

use bayou_diagnostic::span::Span;
use bayou_diagnostic::{Diagnostic, Snippet};

use crate::diagnostic::Diagnostics;
use crate::ir::token::{Keyword, Token, TokenKind};
use crate::ir::Interner;

#[derive(thiserror::Error, Debug)]
pub enum LexerError {
    #[error("unexpected character {0:?}")]
    UnexpectedChar(char),

    #[error("integer overflow")]
    IntegerOverflow,

    #[error("digit {digit:?} is invalid for base {base}")]
    IntegerDigitWrongBase { base: u32, digit: char },
}

pub type LexerResult<T> = Result<T, LexerError>;

pub struct Lexer<'sess> {
    interner: Interner,
    diagnostics: Diagnostics,

    all: &'sess str,
    chars: Chars<'sess>,

    token_start: usize,

    current: Option<Token>,
}

impl<'sess> Lexer<'sess> {
    pub fn new(source: &'sess str) -> Self {
        let mut lexer = Self {
            // TODO: pass in interner and diagnostics
            interner: Interner::new(),
            diagnostics: Diagnostics::default(),

            all: source,
            chars: source.chars(),

            token_start: 0,

            current: None,
        };
        lexer.current = lexer.lex_token();
        lexer
    }

    pub fn finish(self) -> (Interner, Diagnostics) {
        (self.interner, self.diagnostics)
    }

    pub fn lex_token(&mut self) -> Option<Token> {
        loop {
            macro_rules! try_lex {
                ($e:expr) => {{
                    match $e {
                        Ok(token) => token,
                        Err(err) => {
                            self.report_error(err);
                            continue;
                        }
                    }
                }};
            }

            self.token_start = self.byte_pos();

            let kind = match self.chars.next()? {
                // comment
                '/' if self.chars.eat('/') => {
                    while !matches!(self.chars.next(), Some('\n') | None) {}
                    continue;
                }

                ch if ch.is_ascii_whitespace() => continue,

                '{' => TokenKind::LBrace,
                '}' => TokenKind::RBrace,
                '(' => TokenKind::LParen,
                ')' => TokenKind::RParen,
                ';' => TokenKind::Semicolon,

                '+' => TokenKind::Add,
                '-' => TokenKind::Sub,
                '*' => TokenKind::Mul,
                '/' => TokenKind::Div,
                '%' => TokenKind::Mod,

                '&' => TokenKind::BitwiseAnd,
                '|' => TokenKind::BitwiseOr,
                '^' => TokenKind::BitwiseXor,
                '~' => TokenKind::BitwiseInvert,

                '0' if self.chars.eat('x') => try_lex!(self.lex_integer(0, 16)),
                '0' if self.chars.eat('o') => try_lex!(self.lex_integer(0, 8)),
                '0' if self.chars.eat('b') => try_lex!(self.lex_integer(0, 2)),

                ch @ '0'..='9' => try_lex!(self.lex_integer(ch as u64 - 48, 10)),

                ch if is_ident_start(ch) => self.lex_alpha(),

                ch => {
                    self.report_error(LexerError::UnexpectedChar(ch));
                    continue;
                }
            };

            let token = Token {
                kind,
                span: Span::new(self.token_start, self.byte_pos()),
            };

            return Some(token);
        }
    }

    fn lex_integer(&mut self, start: u64, base: u32) -> LexerResult<TokenKind> {
        let mut n = start;

        while let Some(ch @ ('0'..='9' | '_')) = self.chars.peek() {
            self.chars.next();

            if ch == '_' {
                continue;
            }

            let digit = ch
                .to_digit(base)
                .ok_or(LexerError::IntegerDigitWrongBase { base, digit: ch })?;

            n = n
                .checked_mul(base as u64)
                .ok_or(LexerError::IntegerOverflow)?;

            n = n
                .checked_add(digit as u64)
                .ok_or(LexerError::IntegerOverflow)?;
        }

        Ok(TokenKind::Integer(n))
    }

    fn lex_alpha(&mut self) -> TokenKind {
        while matches!(self.chars.peek(), Some(ch) if is_ident(ch)) {
            self.chars.next();
        }

        let s = &self.all[self.token_start..self.byte_pos()];

        match s {
            "int" => TokenKind::Keyword(Keyword::Int),
            "return" => TokenKind::Keyword(Keyword::Return),
            _ => {
                let interned = self.interner.get_or_intern(s);
                TokenKind::Identifier(interned)
            }
        }
    }

    fn byte_pos(&self) -> usize {
        self.all.len() - self.chars.as_str().len()
    }

    fn report_error(&mut self, error: LexerError) {
        let span = Span::new(self.token_start, self.byte_pos());
        self.diagnostics.report(
            Diagnostic::error()
                .with_message(error.to_string())
                .with_snippet(Snippet::primary("this token", 0, span)),
        );
    }
}

impl Iterator for Lexer<'_> {
    type Item = Token;

    fn next(&mut self) -> Option<Self::Item> {
        let token = self.current.take();
        self.current = self.lex_token();
        token
    }
}

impl Peek for Lexer<'_> {
    fn peek(&self) -> Option<Self::Item> {
        self.current
    }
}

pub trait Peek: Iterator {
    fn peek(&self) -> Option<Self::Item>;

    fn eat<P>(&mut self, pat: P) -> bool
    where
        Self::Item: PartialEq<P>,
    {
        match self.peek() {
            Some(item) if item == pat => {
                self.next();
                true
            }
            _ => false,
        }
    }

    fn at_end(&self) -> bool {
        self.peek().is_none()
    }
}

impl Peek for Chars<'_> {
    fn peek(&self) -> Option<Self::Item> {
        self.clone().next()
    }
}

fn is_ident_start(ch: char) -> bool {
    ch.is_ascii_alphabetic() || ch == '_'
}

fn is_ident(ch: char) -> bool {
    ch.is_ascii_alphanumeric() || ch == '_'
}
