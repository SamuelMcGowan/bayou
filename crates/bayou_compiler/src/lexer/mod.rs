#[cfg(test)]
mod tests;

use std::str::Chars;

use bayou_diagnostic::span::Span;

use crate::ir::token::{Keyword, Token, TokenKind};
use crate::ir::Interner;
use crate::utils::peek::Peek;

#[derive(serde::Serialize)]
pub struct LexerError {
    pub kind: LexerErrorKind,
    pub span: Span,
}

#[derive(serde::Serialize, thiserror::Error, Debug)]
pub enum LexerErrorKind {
    #[error("unexpected character {0:?}")]
    UnexpectedChar(char),

    #[error("integer overflow")]
    IntegerOverflow,

    #[error("digit {digit:?} is invalid for base {base}")]
    IntegerDigitWrongBase { base: u32, digit: char },
}

pub type LexerResult<T> = Result<T, LexerErrorKind>;

pub struct Lexer<'sess> {
    interner: &'sess mut Interner,
    errors: Vec<LexerError>,

    all: &'sess str,
    chars: Chars<'sess>,

    token_start: usize,
}

impl<'sess> Lexer<'sess> {
    pub fn new(source: &'sess str, interner: &'sess mut Interner) -> Self {
        Self {
            interner,
            errors: vec![],

            all: source,
            chars: source.chars(),

            token_start: 0,
        }
    }

    pub fn lex(mut self) -> (TokenIter, Vec<LexerError>) {
        let mut tokens = vec![];
        while let Some(token) = self.lex_token() {
            tokens.push(token);
        }

        let iter = TokenIter {
            tokens: tokens.into_iter(),
            prev_span: Span::empty(0),
            eof_span: Span::empty(self.chars.as_str().len()),
        };

        (iter, self.errors)
    }

    fn lex_token(&mut self) -> Option<Token> {
        loop {
            macro_rules! try_lex {
                ($e:expr) => {{
                    match $e {
                        Ok(token) => token,
                        Err(err) => {
                            self.report_error(err);
                            continue;
                        }
                    }
                }};
            }

            self.token_start = self.byte_pos();

            let kind = match self.chars.next()? {
                // comment
                '/' if self.chars.eat('/') => {
                    while !matches!(self.chars.next(), Some('\n') | None) {}
                    continue;
                }

                ch if ch.is_ascii_whitespace() => continue,

                '{' => TokenKind::LBrace,
                '}' => TokenKind::RBrace,
                '(' => TokenKind::LParen,
                ')' => TokenKind::RParen,
                ';' => TokenKind::Semicolon,
                '!' => TokenKind::Bang, // will break `!=`
                '-' if self.chars.eat('>') => TokenKind::Arrow,

                '+' => TokenKind::Add,
                '-' => TokenKind::Sub,
                '*' => TokenKind::Mul,
                '/' => TokenKind::Div,
                '%' => TokenKind::Mod,

                '=' => TokenKind::Assign,

                '&' => TokenKind::BitwiseAnd,
                '|' => TokenKind::BitwiseOr,
                '^' => TokenKind::BitwiseXor,
                '~' => TokenKind::BitwiseInvert,

                '0' if self.chars.eat('x') => try_lex!(self.lex_integer(0, 16)),
                '0' if self.chars.eat('o') => try_lex!(self.lex_integer(0, 8)),
                '0' if self.chars.eat('b') => try_lex!(self.lex_integer(0, 2)),

                ch @ '0'..='9' => try_lex!(self.lex_integer(ch as i64 - 48, 10)),

                ch if is_ident_start(ch) => self.lex_alpha(),

                ch => {
                    self.report_error(LexerErrorKind::UnexpectedChar(ch));
                    continue;
                }
            };

            let token = Token {
                kind,
                span: Span::new(self.token_start, self.byte_pos()),
            };

            return Some(token);
        }
    }

    fn lex_integer(&mut self, start: i64, base: u32) -> LexerResult<TokenKind> {
        let mut n = Some(start);

        while let Some(ch @ ('0'..='9' | 'a'..='f' | 'A'..='F' | '_')) = self.chars.peek() {
            self.chars.next();

            if ch == '_' {
                continue;
            }

            let digit = ch
                .to_digit(base)
                .ok_or(LexerErrorKind::IntegerDigitWrongBase { base, digit: ch })?;

            n = n.and_then(|n| n.checked_mul(base as i64));
            n = n.and_then(|n| n.checked_add(digit as i64));
        }

        n.map(TokenKind::Integer)
            .ok_or(LexerErrorKind::IntegerOverflow)
    }

    fn lex_alpha(&mut self) -> TokenKind {
        while matches!(self.chars.peek(), Some(ch) if is_ident(ch)) {
            self.chars.next();
        }

        let s = &self.all[self.token_start..self.byte_pos()];

        match s {
            "func" => TokenKind::Keyword(Keyword::Func),
            "return" => TokenKind::Keyword(Keyword::Return),
            "let" => TokenKind::Keyword(Keyword::Let),
            "int" => TokenKind::Keyword(Keyword::Int),
            "void" => TokenKind::Keyword(Keyword::Void),
            _ => {
                let interned = self.interner.get_or_intern(s);
                TokenKind::Identifier(interned)
            }
        }
    }

    fn byte_pos(&self) -> usize {
        self.all.len() - self.chars.as_str().len()
    }

    fn report_error(&mut self, kind: LexerErrorKind) {
        let span = Span::new(self.token_start, self.byte_pos());
        self.errors.push(LexerError { kind, span });
    }
}

fn is_ident_start(ch: char) -> bool {
    ch.is_ascii_alphabetic() || ch == '_'
}

fn is_ident(ch: char) -> bool {
    ch.is_ascii_alphanumeric() || ch == '_'
}

pub struct TokenIter {
    tokens: std::vec::IntoIter<Token>,
    prev_span: Span,
    eof_span: Span,
}

impl TokenIter {
    pub fn prev_span(&self) -> Span {
        self.prev_span
    }

    pub fn peek_span(&self) -> Span {
        self.peek().map(|t| t.span).unwrap_or(self.eof_span)
    }

    pub fn eof_span(&self) -> Span {
        self.eof_span
    }
}

impl Iterator for TokenIter {
    type Item = Token;

    fn next(&mut self) -> Option<Self::Item> {
        let token = self.tokens.next()?;
        self.prev_span = token.span;
        Some(token)
    }
}

impl Peek for TokenIter {
    fn peek(&self) -> Option<Self::Item> {
        self.tokens.as_slice().first().copied()
    }
}
