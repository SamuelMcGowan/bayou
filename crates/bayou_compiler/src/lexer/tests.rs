use super::{Lexer, LexerError};
use crate::ir::token::Token;
use crate::ir::Interner;
use crate::utils::assert_yaml_snapshot_with_source;

fn lex(source: &str) -> (Vec<Token>, Vec<LexerError>) {
    let mut interner = Interner::new();

    let lexer = Lexer::new(source, &mut interner);
    let (tokens, lexer_errors) = lexer.lex();

    (tokens.collect(), lexer_errors)
}

macro_rules! assert_lex {
    ($source:expr) => {{
        let source = $source;
        assert_yaml_snapshot_with_source!(source => lex(source));
    }};
}

#[test]
fn integer() {
    assert_lex!("100");
}

#[test]
fn integer_overflow() {
    assert_lex!("100000000000000000000");
}

#[test]
fn newlines() {
    assert_lex!("\nfunc\nmain\n(\n)\n->\nint\n{\nreturn\n0\n;\n}");
}

#[test]
fn no_newlines() {
    assert_lex!("func main()->int{return 0;}");
}

#[test]
fn spaces() {
    assert_lex!("  func  main  (  )  ->  int  {  return  0  ;  }");
}
