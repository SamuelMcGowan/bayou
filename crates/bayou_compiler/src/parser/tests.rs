use super::{ParseError, Parser};
use crate::ir::ast::Module;
use crate::ir::Interner;
use crate::lexer::Lexer;
use crate::utils::assert_yaml_snapshot_with_source;

fn parse(source: &str) -> (Module, Vec<ParseError>) {
    let mut interner = Interner::new();

    let lexer = Lexer::new(source, &mut interner);
    let (tokens, lexer_errors) = lexer.lex();

    assert!(lexer_errors.is_empty(), "lexer errors in parser tests");

    let parser = Parser::new(tokens);
    parser.parse()
}

macro_rules! assert_parse {
    ($source:expr) => {{
        let source = $source;
        assert_yaml_snapshot_with_source!(source => parse(source));
    }};
}

#[test]
fn return_integer() {
    assert_parse!("func main() -> i64 { return 0; }");
}

#[test]
fn missing_paren() {
    assert_parse!("func main( { return 0; }");
}

#[test]
fn no_return_value() {
    assert_parse!("func main() { return; }");
}

#[test]
fn no_brace() {
    assert_parse!("func main() -> i64 { return 0;");
}

#[test]
fn no_semicolon() {
    assert_parse!("func main() -> i64 { return 0 }");
}

#[test]
fn no_semicolon_or_return_value() {
    assert_parse!("func main() -> i64 { return }");
}

#[test]
fn no_space() {
    assert_parse!("func main() -> i64 { return0; }");
}

#[test]
fn wrong_case() {
    assert_parse!("func main() -> i64 { RETURN 0; }");
}
