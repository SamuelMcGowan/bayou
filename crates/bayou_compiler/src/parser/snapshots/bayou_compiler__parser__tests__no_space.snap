---
source: crates/bayou_compiler/src/parser/tests.rs
info: "func main() -> int { return0; }"
---
- items:
    - FuncDecl:
        name:
          ident: 1
          span:
            start: 5
            end: 9
        ret_ty: I64
        statements:
          - ParseError
- - expected: a statement
    span:
      start: 21
      end: 28

